src/cli/self_update.rs

```diff
@@ -86,9 +89,53 @@ pub const NEVER_SELF_UPDATE: bool = true;
 #[cfg(not(feature = "no-self-update"))]
 pub const NEVER_SELF_UPDATE: bool = false;
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum SelfUpdateMode {
+    Enable,
+    Disable,
+    CheckOnly,
+}
+
+impl SelfUpdateMode {
+    pub fn modes() -> &'static [&'static str] {
+        &["enable", "disable", "check-only"]
+    }
+
+    pub fn default_mode() -> &'static str {
+        "enable"
+    }
+}
+
+impl FromStr for SelfUpdateMode {
+    type Err = anyhow::Error;
+
+    fn from_str(mode: &str) -> Result<Self> {
+        match mode {
+            "enable" => Ok(Self::Enable),
+            "disable" => Ok(Self::Disable),
+            "check-only" => Ok(Self::CheckOnly),
+            _ => Err(anyhow!(format!(
+                "unknown self update mode: '{}'; valid modes are {}",
+                mode,
+                valid_self_update_modes(),
+            ))),
+        }
+    }
+}
+
+impl ToString for SelfUpdateMode {
+    fn to_string(&self) -> String {
+        let modes = Self::modes();
+        match self {
+            SelfUpdateMode::Enable => modes[0].to_string(),
+            SelfUpdateMode::Disable => modes[1].to_string(),
+            SelfUpdateMode::CheckOnly => modes[2].to_string(),
```

Comment:
I don't think the modes array here adds anything: you have connaisence of value between the ordering of the array and the values in the lookup, vs connaisence of value between an inline string here and a string in FromStr; string interning should optimise the binary size the same either way, and so it becomes what is easier to read: 

```
match self {
   SelfUpdateMode::Enable => "enable",
   SelfUpdateMode::Disable => "disable"
   SelfUpdateMode::CheckOnly => "check-only"
}.into()
```

would be my preference - but its ok if you don't choose to change it; just making a note in passing.